Cast screen kms

yhe39 opened the pull request.

sysopenci commented:
Improper Commit Message
Tracked on not found in commit message,
    make sure Tracked-On: Jira-ticket  is present.

sysopenci commented:
Improper Commit Message
Tracked on not found in commit message,
    make sure Tracked-On: Jira-ticket  is present.

yhe39 committed:
Add probe new added render node

Open the first available ivshm node.
If it is sancout buffer with certain resolution,
use the ivshm node to allocate buffer.

Tracked-On: OAM-122966
Signed-off-by: He, Yue <yue.he@intel.com>

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -27,6 +27,9 @@
 #define DRM_CARD_NODE_START 0
 
+#define IVSH_WIDTH 1920
+#define IVSH_HEIGHT 1080
```
The width/height is too common, just 1080p display mode. When connect with 1080p monitor, it should lead to some problem.

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +112,103 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+	char const *str = "%s/renderD%d";
+	const char *undesired[2] = { "vgem", nullptr };
+	uint32_t min_node = 128;
+	uint32_t max_node = (min_node + drv_num + 2);
```
2 is magic number, it is better define one macro for it, for example:
#define IVSHMEME_DEVICE_NUM 2

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +112,103 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+	uint32_t gpu_grp_type = 0;
+
+	availabe_node = 0;
+	for (uint32_t i = min_node; i < max_node; i++) {
```
we only need check new added device, then can we just use
for (uint32_t i = 128 + drv_num; i < 128 + drv_num + IVSHMEME_DEVICE_NUM ; i++) {

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +112,103 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+		}
+
+		for (j = 0; j < ARRAY_SIZE(undesired); j++) {
+			if (undesired[j] && !strcmp(version->name, undesired[j])) {
```
we only compare with vgem, why need a for loop here?
what's the meaning of vgem here?

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +112,103 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+		drmFreeVersion(version);
+	}
+
+	// have the ivshemem node
```
can we call ioctl VIRTGPU_PARAM_QUERY_DEV to double confirm the new added device is ivshmem?

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -331,7 +453,14 @@ bool cros_gralloc_driver::get_resolved_format_and_use_flags(
 	struct driver *drv = is_video_format(descriptor) ? drv_video_ : drv_render_;
-	if ((descriptor->use_flags & BO_USE_SCANOUT) && !(is_video_format(descriptor)))
+	if (!drv_ivshmem_) {
+		if (reload()) {
```
reload will be always called if ACRN didn't expose ivshmem device to Android, then the overload is too heavy.
Can we just check the special width/height, then reload?

yhe39 commented:
@feijiang1 @phreer @chenyanxzhu  the pr is updated, please check.

phreer commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +114,96 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+		}
+		gpu_grp_type = GPU_WITH_IVSHMEM;
+
+		if (drv_ivshmem_) {
```
Looks like this check is redundant since it was already checked in line 182 already.

phreer commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -177,7 +277,10 @@ cros_gralloc_driver::cros_gralloc_driver()
 		if (!strcmp(version->name, "virtio_gpu")) {
-			virtio_node_idx = availabe_node;
+			if (virtio_node_idx == -1)
+				virtio_node_idx = availabe_node;
```
We are assuming that the first virtio-GPU device would always be virtio-pci device, and the others are virtio-ivshmem devices, right? Why don't we call `drv_virtgpu_is_ivshm()` to check it?

yhe39 replied:
The drv_virtgpu_is_ivshm() will call success after the drv init, in this step, the driver is not create/init, couldn't get the the result.

phreer commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -331,7 +448,10 @@ bool cros_gralloc_driver::get_resolved_format_and_use_flags(
 
 	struct driver *drv = is_video_format(descriptor) ? drv_video_ : drv_render_;
-	if ((descriptor->use_flags & BO_USE_SCANOUT) && !(is_video_format(descriptor)))
+	if (drv_ivshmem_ && (descriptor->use_flags & BO_USE_SCANOUT) &&
```
Policy of selecting driver getting complicated, would be better to introduce a new function for it.

phreer commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +114,96 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+
+	// have the ivshemem node
+	if ((availabe_node > 0) && !drv_ivshmem_) {
+		drv_num_ = drv_num_ + availabe_node;
```
If I understand correctly, you are searching for the render node with largest id, but it seems the result should't be calculated as this. If there is an empty slot, the result is incorrect. For example, if we have renderD128, renderD129, and renderD131 (somehow the renderD130 node is not present), you will try to open renderD130.

yhe39 replied:
i don't understand in which case there is empty slot.

yhe39 replied:
change accordingly, please check.

chenyanxzhu commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -262,6 +367,15 @@ cros_gralloc_driver::cros_gralloc_driver()
+					ivshm_node_idx);
+				close(node_fd[ivshm_node_idx]);
+			} else
+				gpu_grp_type = GPU_WITH_IVSHMEM;
```
The other value may be overlapped here, which will impact i915 code below. And GPU_WITH_IVSHMEM is not used anywhere.
		if (((drv->gpu_grp_type == TWO_GPU_IGPU_DGPU) ||
		    (drv->gpu_grp_type == THREE_GPU_IGPU_VIRTIO_DGPU))
			&& (i915->graphics_version >= 125))
			return 0;

yhe39 replied:
Just thought that the gpu scenario is more complicate, i add some gpu type, please check.

chenyanxzhu commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +114,96 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+			return -ENODEV;
+		}
+
+		if (strcmp(version->name, "virtio_gpu")) {
```
Code from 174 to 202, may be included in the previous for loop in line 148? And code will look like simpler. And don't forget drmFreeVersion.

chenyanxzhu commented on drv_priv.h:
```diff
@@ -65,7 +65,11 @@ enum CIV_GPU_TYPE {
+	FOUR_GPU_IGPU_VIRTIO_DGPU_IVHM,
+	FOUR_GPU_IGPU_VIRTIO_TWO_IVHM,
+	FIVE_GPU_IGPU_VIRTIO_DGPU_TWO_IVHM
 };
```
I feel like the code here is getting a little complicated. Maybe use another high bit flag to distinguish the ivsh/non-ivsh status if that is needed, like: GPU_WITH_IVSHMEM = (1<<16) ? Also, "gpu_grp_type" should make a member of cros_gralloc_driver, as in func "reload()" it's not initialized.

yhe39 replied:
Yes, you are right. So far the gpu_grp_type is only used to identify whether it has dual gpu, so i modify this logic, please check.
Thanks！

phreer replied:
> I feel like the code here is getting a little complicated. Maybe use another high bit flag to distinguish the ivsh/non-ivsh status if that is needed, like: GPU_WITH_IVSHMEM = (1<<16) ? Also, "gpu_grp_type" should make a member of cros_gralloc_driver, as in func "reload()" it's not initialized.

Agree, making `gpu_grp_type` a member of `cros_gralloc_driver` would be better. Otherwise we have a few instances of `gpu_grp_type` and should make them consistent.

yhe39 commented:
@feijiang1 @phreer @chenyanxzhu the pr is updated, please check.

feijiang1 commented on i915.c:
```diff
@@ -443,9 +443,7 @@ static int i915_add_combinations(struct driver *drv)
-		if (((drv->gpu_grp_type == TWO_GPU_IGPU_DGPU) ||
-		    (drv->gpu_grp_type == THREE_GPU_IGPU_VIRTIO_DGPU))
-			&& (i915->graphics_version >= 125))
+		if ((drv->gpu_grp_type && GPU_TYPE_DUAL_IGPU_DGPU) && (i915->graphics_version >= 125))
```
&& should be &

feijiang1 commented on i915.c:
```diff
@@ -476,8 +474,7 @@ static int i915_add_combinations(struct driver *drv)
 							    .modifier = I915_FORMAT_MOD_Y_TILED };
-		if ((drv->gpu_grp_type == TWO_GPU_IGPU_DGPU) ||
-		    (drv->gpu_grp_type == THREE_GPU_IGPU_VIRTIO_DGPU)) {
+		if (drv->gpu_grp_type && GPU_TYPE_DUAL_IGPU_DGPU) {
```
&& should be &

chenyanxzhu commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +114,86 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+			return -ENODEV;
+		}
+
+		gpu_grp_type = GPU_TYPE_NORMAL;
```
This assignment here still will overlap previous GPU_TYPE_DUAL_IGPU_DGPU flag. Use "|=" maybe better? 'reload' is executed after 'cros_gralloc_driver()', that's why i think 'gpu_grp_type' should make a member of 'cros_gralloc_driver'

feijiang1 commented on cros_gralloc/cros_gralloc_driver.h:
```diff
@@ -90,6 +93,9 @@ class cros_gralloc_driver
 	struct driver *drv_kms_ = nullptr;
+	struct driver *drv_ivshmem_ = nullptr;
+	uint32_t drv_num_ = 0;
+	uint64_t gpu_grp_type = GPU_TYPE_NORMAL;
```
gpu_grp_type should be changed to gpu_grp_type_

feijiang1 commented on cros_gralloc/cros_gralloc_driver.cc:
```diff
@@ -109,6 +114,83 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
+			return -ENODEV;
+		}
+
+		fd = open(node, O_RDWR, 0);
```
need check fd, here, upper log
drv_loge("Open new added node fail\n");
need be moved here

yhe39 replied:
in the line 136, checked the fd, the availabe_node calculate the available node, in line 152, availabe_node > 0 means has the available node, so the fd should be the workable fd.
I add the checks before, Chenyan comment that the checks are done above, so i remove the checks.

chenyanxzhu replied:
Actually what i mean is that u can push this bloc code(line 163-190) into the for loop (line 145), and rm line 152-162. Just for reference. Also 'availabe_node' can be removed if change this way.

feijiang1 replied:
agree, suggest to combine upper code to remove some duplicated code. Also please help add some comments here, for example, the second ivshmem device is for our screen cast purpose.
Also please add enough introduction in the commit description.

yhe39 replied:
Patch updated, i made changes to use the first available ivshm node to screen cast.
Please check.

feijiang1 reviewed (approved):
LGTM

feijiang1 reviewed (approved):
LGTM

sysopenci commented:
Android CI has started Engineering Build for this issue ,Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
Android CI has started MERGE Build for this pr ,Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
Android CI has completed MERGE Build for this pr, build is SUCCESS. Please check the linked Tracked-On issue/Android CI Web  for more details. For Binaries: /cactus-absp-or-local/celadon_umr0_master-merge/94
